pyOCD/gdbserver/gdbserver.py

```diff
@@ -329,12 +338,27 @@ def run(self):
                     break
 
                 if self.packet_io.interrupt_event.isSet():
-                    logging.debug("Got unexpected ctrl-c, ignoring")
+                    if self.non_stop:
+                        self.target.halt()
+                        self.is_target_running = False
+                        self.sendStopNotification()
+                    else:
+                        logging.debug("Got unexpected ctrl-c, ignoring")
                     self.packet_io.interrupt_event.clear()
 
+                if self.non_stop and self.is_target_running:
+                    try:
+                        if self.target.getState() == TARGET_HALTED:
+                            logging.debug("state halted")
+                            self.is_target_running = False
+                            self.sendStopNotification()
+                    except Exception as e:
+                        logging.debug("Unexpected exception: %s", e)
+                        traceback.print_exc()
+
                 # read command
                 try:
-                    packet = self.packet_io.receive()
+                    packet = self.packet_io.receive(block=not self.non_stop)
```

Comment:
If I block here, then we can't check for ctrl-c events. This could probably be reorganized to pass a timeout into `receive()` instead of setting non-blocking and then calling `sleep()` a few lines down.  `GDBServer.syscall()` also uses the same pattern and could be similarly changed. What is your preference?